### PR TITLE
ARROW-10089: [R] inject base class for Array, ChunkedArray and Scalar

### PR DESCRIPTION
**Pull request by nealrichardson**
The impetus for this change was to get S3 Ops method dispatch to work for Array == Scalar. This is tested in the third commit.

### CODE REVIEW

File: r/R/table.R [hunk 1]
```diff
@@ -248,39 +203,9 @@ Table$create <- function(..., schema = NULL) {
...
-`names<-.Table` <- function(x, value) x$RenameColumns(value)
-
-#' @export
-`[.Table` <- `[.RecordBatch`
-
-#' @export
-`[[.Table` <- `[[.RecordBatch`
-
 #' @export
 `[[<-.Table` <- function(x, i, value) {
```
Comment (ianmcook) [inline r/R/table.R:210]: Could you add a TODO/comment here describing why these have not yet been changed to be methods of the base class `ArrowTabular`? And another one where `$<-.Table` is defined

**Comment by ianmcook** [comment]
@nealrichardson could you please confirm that this is correct?
- `RecordBatch` and `Table` now inherit the synthetic class `ArrowTabular`
- `Array`, `ChunkedArray`, and `Scalar` now inherit the synthetic class `ArrowDatum`
These synthetic classes do not exist in the C++ class hierarchy; they are used here because they simplify S3 method dispatch.

File: r/R/table.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/table.R]: I ended up moving them to ArrowTabular and left `stop(TODO)` messages in the missing RecordBatch methods.

**Comment by nealrichardson** [comment]
> @nealrichardson could you please confirm that this is correct?
> 
> * `RecordBatch` and `Table` now inherit the synthetic class `ArrowTabular`
> * `Array`, `ChunkedArray`, and `Scalar` now inherit the synthetic class `ArrowDatum`
> * These synthetic classes do not exist in the C++ class hierarchy; they are used here because they simplify S3 method dispatch.

Correct, and I expanded on the comments about that in the code.

**Pull request closed without merging**